File: java/program/Computer.java
import java.util.ArrayList;
import java.util.List;

public class Computer {
    private String type;
    private OperatingSystem os;
    private Memory memory;
    private List<HardDrive> hardDrives;  
    private List<Application> applications;

    // Constructor
    public Computer(String type, OperatingSystem os, Memory memory) {
        this.type = type;
        this.os = os;
        this.memory = memory;
        this.hardDrives = new ArrayList<>();
        this.applications = new ArrayList<>();
    }

    // Add & Remove Hard Drives
    public void addHardDrive(HardDrive hd) {
        hardDrives.add(hd);
    }

    public void removeHardDrive(String name) {
        hardDrives.removeIf(hd -> hd.getName().equals(name));
    }

    // Add & Remove Applications
    public void addApplication(Application app) {
        applications.add(app);
    }

    public void removeApplication(String name) {
        applications.removeIf(app -> app.getName().equals(name));
    }

    // Getters
    public String getType() {
        return type;
    }

    public OperatingSystem getOS() {
        return os;
    }

    public Memory getMemory() {
        return memory;
    }

    public List<HardDrive> getHardDrives() {
        return hardDrives;
    }

    public List<Application> getApplications() {
        return applications;
    }

    // Setters
    public void setType(String type) {
        this.type = type;
    }

    public void setOS(OperatingSystem os) {
        this.os = os;
    }

    public void setMemory(Memory memory) {
        this.memory = memory;
    }

    
    public void printInfo() {
        System.out.println("Computer Type: " + type);
        System.out.println("OS: " + os.getName());
        System.out.println("Memory: " + memory.getSize() + "GB");

        System.out.println("Hard Drives:");
        for (HardDrive hd : hardDrives) {
            System.out.println(" - " + hd.getName() + " (" + hd.getCapacity() + " GB)");
        }

        System.out.println("Applications:");
        for (Application app : applications) {
            System.out.println(" - " + app.getName() + " (" + app.getCategory() + ")");
        }
        System.out.println("~.~.~\n");
    }
}

File: cpp/Computer.cpp
#pragma once
#include <vector>
#include <algorithm>

#include "OperatingSystem.cpp"
#include "Memory.cpp"
#include "HardDrive.cpp"
#include "Application.cpp"

using namespace std;

class Computer {
private:
    string type;
    OperatingSystem os;
    Memory mem;
    vector<HardDrive> hardDrives;  // Menggunakan vector untuk banyak hard drive
    vector<Application> applications;

    public:
    // Constructor with references to avoid unnecessary copies
    Computer(const string& type, const OperatingSystem& os, const Memory& mem)
        : type(type), os(os), mem(mem) {}  // Initialize members

    // Add functions to handle adding/removing hard drives and applications
    void addHardDrive(const HardDrive& hd) { hardDrives.push_back(hd); }
    void removeHardDrive(const string& name) {
        hardDrives.erase(remove_if(hardDrives.begin(), hardDrives.end(),
            [&](const HardDrive& hd) { return hd.getName() == name; }),
            hardDrives.end());
    }

    void addApplication(const Application& app) { applications.push_back(app); }
    void removeApplication(const string& name) {
        applications.erase(remove_if(applications.begin(), applications.end(),
            [&](const Application& app) { return app.getName() == name; }),
            applications.end());
    }

    // Getters
    const string& getType() const { return type; }
    const OperatingSystem& getOS() const { return os; }
    const Memory& getMemory() const { return mem; }
    const vector<HardDrive>& getHardDrives() const { return hardDrives; }
    const vector<Application>& getApplications() const { return applications; }


    // Setter
    void setType(string t) { type = t; }
    void setOS(OperatingSystem o) { os = o; }
    void setMemory(Memory m) { mem = m; }
    
    // void addHardDrive(HardDrive hd) { hardDrives.push_back(hd); }
    // void removeHardDrive(int index) {
    //     if (index >= 0 && index < hardDrives.size()) {
    //         hardDrives.erase(hardDrives.begin() + index);
    //     }
    // }

    // void addApplication(Application app) { applications.push_back(app); }
    // void removeApplication(int index) {
    //     if (index >= 0 && index < applications.size()) {
    //         applications.erase(applications.begin() + index);
    //     }
    // }
};
